fuse_graphs/src/hash_graph.cpp

```diff
@@ -53,6 +53,8 @@ namespace fuse_graphs
 HashGraph::HashGraph(const HashGraphParams& params) :
   problem_options_(params.problem_options)
 {
+  // We don't want Ceres to take ownership of the loss functions
+  problem_options_.loss_function_ownership = ceres::Ownership::DO_NOT_TAKE_OWNERSHIP;
```

Comment:
This is important because we recreate the problem every time the graph is optimized or we get the covariance from the solver. And even if we didn't recreate the problem, now we need to guarantee the `ceres::LossFunction`s don't get destroyed until the `fuse_core::Loss` wrapper that handles it is destroyed, which happens when the `fuse_core::Constraint` that has that loss function is not used anymore, e.g. marginalized out.

Reply:
I think using `DO_NOT_TAKE_OWNERSHIP` would allow some optimizations gains. That could be done in a future PR.

I've done some changes that now make the loss functions behave as the constraints, so ownership is on the code that calls `lossFunction()`.

I also added this to the `Loss` base class:
``` c++
  static constexpr ceres::Ownership Ownership =
      ceres::Ownership::TAKE_OWNERSHIP;  //<! The ownership of the ceres::LossFunction* returned by lossFunction()
```

This way the code that uses it can check the ownership that's being used and act accordingly. I've updated all the code and tests to take this into account. This means that switching to `DO_NOT_TAKE_OWNERSHIP` only requires a change here, in principle.